### PR TITLE
Use HashSet for machine block update and limit recursion depth

### PR DESCRIPTION
**Pull request by Mogball**
As per https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/5906

* Changed `ArrayList` to `HashSet`
* Halted update recursion on cables, fluid pipes, and item pipes
* Set a max recursion depth, which is hopefully large enough for anything but the mega machines
* Took the update out of a thread

From my testing, I found that multiblock machines would still form correctly and detect hatches, etc.

### CODE REVIEW

File: src/main/java/gregtech/api/threads/GT_Runnable_MachineBlockUpdate.java [hunk 1]
```diff
@@ -1,43 +1,79 @@
...
+          return true;
+        IMetaTileEntity tMetaTileEntity = ((IGregTechTileEntity) aTileEntity).getMetaTileEntity();
+        return
+            !(tMetaTileEntity instanceof GT_MetaPipeEntity_Cable) &&
+            !(tMetaTileEntity instanceof GT_MetaPipeEntity_Fluid) &&
+            !(tMetaTileEntity instanceof GT_MetaPipeEntity_Item);
+    }
+
+    private void stepToUpdateMachine(World aWorld, int aX, int aY, int aZ) {
+        mVisited.add(new ChunkPosition(aX, aY, aZ));
```
Comment (mitchej123) [inline src/main/java/gregtech/api/threads/GT_Runnable_MachineBlockUpdate.java:53]: Rather than checking `mVisited.contains`; you could just use if `!mVisiited(add)` -- it'll return false if it's already in the list.

File: src/main/java/gregtech/api/threads/GT_Runnable_MachineBlockUpdate.java [hunk 2]
```diff
@@ -1,43 +1,79 @@
...
+        if (aTileEntity == null)
+          return false;
+
+        // Stop recursion on GregTech cables, item pipes, and fluid pipes
+        if (!(aTileEntity instanceof IGregTechTileEntity))
+          return true;
+        IMetaTileEntity tMetaTileEntity = ((IGregTechTileEntity) aTileEntity).getMetaTileEntity();
+        return
+            !(tMetaTileEntity instanceof GT_MetaPipeEntity_Cable) &&
+            !(tMetaTileEntity instanceof GT_MetaPipeEntity_Fluid) &&
```
Comment (repo-alt) [inline src/main/java/gregtech/api/threads/GT_Runnable_MachineBlockUpdate.java:48]: why not just `instanceof MetaPipeEntity` ?

File: src/main/java/gregtech/api/GregTech_API.java [hunk 3]
```diff
@@ -390,7 +390,7 @@ public static ItemStack getUnificatedOreDictStack(ItemStack aOreStack) {
      */
     public static boolean causeMachineUpdate(World aWorld, int aX, int aY, int aZ) {
         if (!aWorld.isRemote)
-            new Thread(new GT_Runnable_MachineBlockUpdate(aWorld, aX, aY, aZ), "Machine Block Updating").start();
```
Comment (bartimaeusnek) [inline src/main/java/gregtech/api/GregTech_API.java:393]: why was this changed to a sync method instead of a threaded one?

Comment (Mogball) [reply src/main/java/gregtech/api/GregTech_API.java:393]: From @repo-alt's suggestion [here](https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/5906#issuecomment-612632685)

File: src/main/java/gregtech/api/threads/GT_Runnable_MachineBlockUpdate.java (hunk 2 above)
Comment (Mogball) [reply src/main/java/gregtech/api/threads/GT_Runnable_MachineBlockUpdate.java:48]: I saw that frames (`GT_MetaPipeEntity_Frame`) extend `MetaTileEntity` and wanted to be explicit with the cables and item/fluid pipes. Unless excluding everything is okay.

File: src/main/java/gregtech/api/GregTech_API.java (hunk 3 above)
Comment (repo-alt) [reply src/main/java/gregtech/api/GregTech_API.java:393]: because thread spinning is not free (seriously)

File: src/main/java/gregtech/api/threads/GT_Runnable_MachineBlockUpdate.java (hunk 2 above)
Comment (repo-alt) [reply src/main/java/gregtech/api/threads/GT_Runnable_MachineBlockUpdate.java:48]: I see, yes, I missed that frame is a kind of pipe (a bit weird that)

**Comment by Mogball** [comment]
I removed the recursion limit since for really large machines (e.g. mega distillery) a hatch placed on the opposite side of the controller would not trigger an update. I assume this is not acceptable. 

The performance gain of stopping recursion on pipes/cables and switching to `HashSet` are enough for this PR.

**Comment by draknyte1** [comment]
Use debug mode in your IDE, hotswap code on the go. 🤦🏻‍♀️

You can always spin up one thread and keep it alive, no performance hit if it’s just chilling in the background waiting for tasks to be passed to it.

**Comment by Mogball** [comment]
> Use debug mode in your IDE, hotswap code on the go. 🤦🏻‍♀️

Guess I should use an IDE

**Comment by draknyte1** [comment]
Did Dream actually read the two reviews left? 🤦

**Comment by Dream-Master** [comment]
Sorry was to fast with clicking @draknyte1 
Can you make a new one @Mogball with the suggested changes.
Thanks

**Comment by Mogball** [comment]
Okay, I think the remaining discussion was on threaded vs. sync

**Comment by Dream-Master** [comment]
@Mogball yea feel free to open up a new pull. Let me know when it is done and approved. Thanks
